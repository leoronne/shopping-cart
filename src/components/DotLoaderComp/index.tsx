import React from 'react';
import { DotLoader } from 'react-spinners';

interface DotProps {
  loading?: boolean;
  size?: number;
  color?: string;
  defaultText?: any;
}

const DotLoaderComp: React.FC<DotProps> = ({ loading, size, color, defaultText }) => {
  return (
    <>
     {loading ? (
      <DotLoader
        css={`display: block; margin: 0 auto; border-color: red;`}
        size={size}
        color={color}
      />
    ) : (
      defaultText
    )}
    </>
  );
};

export default DotLoaderComp;
