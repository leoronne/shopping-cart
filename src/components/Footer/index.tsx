import React from 'react';

import { Container, GithubIcon } from './styles';

const Footer: React.FC = () => {
  return (
    <Container data-testid="footer-cointainer">
      <p className="text--center">{`Â©${new Date().getFullYear()}. All Rights Reserved.`}</p>
      <a href="https://github.com/leoronne/shopping-cart" target="_blank" rel="noopener noreferrer">
        <GithubIcon />
      </a>
    </Container>
  );
};

export default Footer;
