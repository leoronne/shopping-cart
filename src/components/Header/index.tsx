import React from 'react';
import { Tooltip } from '@material-ui/core';

import logo from '../../assets/img/logo.png';
import avatar from '../../assets/img/avatar.jpg';

import { useStyles } from '../../styles/MaterialUI';
import { Container, Left, Right, NavProfile } from './styles';

const Header: React.FC = () => {
  const classes = useStyles();
  const userName = 'Leonardo Ronne';

  return (
    <Container>
      <Left>
        <a href="/">
          <img src={logo} alt="Shooping Cart" />
        </a>
        <p>Shopping</p>
      </Left>

      <Right>
        <NavProfile>
          <Tooltip title={`Connected as ${userName}`} placement="left" arrow classes={{ tooltip: classes.tooltip }}>
            <a href="https://github.com/leoronne" className="icon-profile-active" target="_blank" rel="noopener noreferrer">
              <img src={avatar} alt={userName} className="avatar-profile" />
              <span>{userName}</span>
            </a>
          </Tooltip>
        </NavProfile>
      </Right>
    </Container>
  );
};

export default Header;
