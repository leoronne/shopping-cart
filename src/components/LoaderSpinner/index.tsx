import React from 'react';
import { CircularProgress } from '@material-ui/core';

import { Container } from './styles';

const Loader: React.FC<{ color?: string }> = ({ color = '#a0101b' }) => {
  return (
    <Container className="loader-container" data-testid="loader-spinner">
      <CircularProgress size={15} style={{ color }} />
    </Container>
  );
};
export default Loader;
