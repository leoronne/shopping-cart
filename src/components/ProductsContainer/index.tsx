import React from 'react';

import { useProducts } from '../../hooks';

import ProductCard from './ProductCard';

import { Container, ProductsListcontainer } from './styles';

const ProductsContainer: React.FC = () => {
  const { products } = useProducts();
  return (
    <Container>
      <ProductsListcontainer>
        {products.map((product) => (
          <ProductCard name={product.name} value={product.price} quantity={product.available} id={product.id} key={product.id} />
        ))}
      </ProductsListcontainer>
    </Container>
  );
};

export default ProductsContainer;
