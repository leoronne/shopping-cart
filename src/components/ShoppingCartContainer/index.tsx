import React, { useState, useCallback } from 'react';
import { CircularProgress } from '@material-ui/core';
import { useSnackbar } from 'notistack';

import formatNumber from '../../utils/formatNumber';

import { useCart, useVouchers } from '../../hooks';

import TotalValues from './TotalValues';
import ProductRow from './ProductRow';

import { Container, ShoppingCart, ProductList, DiscountCode } from './styles';

const ShoppingCartContainer: React.FC = () => {
  const { cartItems, shipping, discount, discountValue, totalItemValue, totalValue, voucherApplied, removeDiscounts, addDiscount, setCartItems } = useCart();
  const { vouchers } = useVouchers();

  const { enqueueSnackbar } = useSnackbar();

  const [voucherCode, setVoucherCode] = useState('');
  const [loadingOrder, setLoadingOrder] = useState(false);

  const handleRemoveDiscounts = useCallback(() => {
    removeDiscounts();
    setVoucherCode('');
  }, [removeDiscounts]);

  const handleDiscounts = useCallback(
    (code?: string) => {
      try {
        if (voucherApplied) enqueueSnackbar('Sorry, you have already applied a voucher code', { variant: 'warning' });
        else {
          const cupom = vouchers.find((voucher) => voucher.code === (code || voucherCode));

          if (!cupom) {
            setVoucherCode('');
            addDiscount({ type: null, value: 0, desc: '' }, false);
            enqueueSnackbar('Sorry, you entered an invalid voucher', { variant: 'error' });
          } else {
            addDiscount(
              {
                type: cupom?.type,
                value: cupom?.type === 'shipping' ? cupom.minValue : cupom.amount,
                desc: `The Voucher qualifies for a ${cupom?.type} discount`,
              },
              true
            );
            enqueueSnackbar('Voucher Applied!', { variant: 'success' });
          }
        }
      } catch (err) {
        // eslint-disable-next-line no-console
        console.error(err.message);
      }
    },
    [voucherCode, voucherApplied, vouchers, addDiscount, enqueueSnackbar]
  );

  const handleSubmitOrder = useCallback(() => {
    setLoadingOrder(true);
    setTimeout(() => {
      enqueueSnackbar('Thank you, we received your order!', { variant: 'success' });
      setCartItems([]);
      handleRemoveDiscounts();
      setLoadingOrder(false);
    }, 1000);
  }, [enqueueSnackbar, setCartItems, handleRemoveDiscounts]);

  return (
    <Container>
      <ShoppingCart>
        <div className="title">
          <p>Cart</p>
        </div>

        <div className="cart-container">
          <ProductList>
            {cartItems.map((item) => {
              return <ProductRow name={item.name} quantity={item.available} id={item.id} key={item.id} total={formatNumber(item.price * item.available)} />;
            })}
          </ProductList>

          <DiscountCode>
            {voucherApplied ? (
              <>
                <input type="text" value={voucherCode} placeholder="Voucher Code" name="voucher" disabled />
                <button type="button" onClick={handleRemoveDiscounts} data-testid="remove-voucher">
                  REMOVE
                </button>
              </>
            ) : (
              <form
                onSubmit={(e: React.FormEvent<HTMLFormElement>) => {
                  e.preventDefault();
                  handleDiscounts();
                }}
              >
                <input
                  type="text"
                  value={voucherCode}
                  placeholder="Voucher Code"
                  name="voucher"
                  onChange={(e) => {
                    setVoucherCode(e.target.value);
                  }}
                  data-testid="voucher-input"
                />
                <button type="submit" disabled={!voucherCode} data-testid="submit-voucher">
                  APPLY
                </button>
              </form>
            )}
          </DiscountCode>
        </div>

        <TotalValues
          items={cartItems}
          values={{
            shipping,
            discount,
            discountValue,
            totalItemValue,
            totalValue,
          }}
        />
      </ShoppingCart>
      <button type="button" disabled={!!(cartItems.length === 0 || loadingOrder)} onClick={handleSubmitOrder} className="checkout-button" data-testid="submit-order">
        {loadingOrder ? (
          <>
            <CircularProgress size={15} style={{ color: '#fff' }} />
          </>
        ) : (
          <span>CHECKOUT</span>
        )}
      </button>
    </Container>
  );
};

export default ShoppingCartContainer;
