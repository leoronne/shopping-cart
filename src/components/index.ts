export { default as LoaderSpinner } from './LoaderSpinner';
export { default as Header } from './Header';
export { default as Footer } from './Footer';
export { default as ProductsContainer } from './ProductsContainer';
export { default as ShoppingCartContainer } from './ShoppingCartContainer';
