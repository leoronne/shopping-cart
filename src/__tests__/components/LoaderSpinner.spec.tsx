import React from 'react';
import { render } from '@testing-library/react';

import { LoaderSpinner } from '../../components';

describe('LoaderSpinner', () => {
  it('should render', async () => {
    const { asFragment } = render(<LoaderSpinner />);

    const html = asFragment();

    expect(html).toMatchSnapshot();
  });
});
