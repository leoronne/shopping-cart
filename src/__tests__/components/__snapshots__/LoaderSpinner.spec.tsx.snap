// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderSpinner should render 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ kpcYzi loader-container"
    data-testid="loader-spinner"
  >
    <div
      class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 15px; height: 15px; color: rgb(160, 16, 27);"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
